### PR TITLE
mitigate high CPU cost of HBHE hit position calculation in EgammaHcalIsolation

### PR DESCRIPTION
**Pull request by slava77**
costs of computing HCAL isolation in egamma has increased significantly after #33666
- [12_0_0_pre3](https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/releases/CMSSW_12_0_0_pre3/11834.21/step3/cpu_endjob/1108) has \~0.14% of reco time in isolation
- [12_0_0_pre4](https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/releases/CMSSW_12_0_0_pre5/11834.21/step3/cpu_endjob/27) has 8.9%
    - [12_1_0_pre4](https://cms-reco-profiling.web.cern.ch/cms-reco-profiling/cgi-bin/igprof-navigator/releases/CMSSW_12_1_0_pre4/11834.21/step3/cpu_endjob/28) has a similar fraction

This PR mitigates the situation by reordering the calls.

However, a better refactoring of the code should be considered in near future. Some options are
- accumulate all needed isolation types in one loop over rechits
- precompute (η,φ) of rechits in a separate producer in a `ValueMap` and use it when dR computations are needed
@wrtabb @SohamBhattacharya

### CODE REVIEW

**Comment by slava77** [comment]
enable profiling

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
in IB igprof tests the perf counts in `EgammaHcalIsolation::getHcalSum` have reduced by about a factor of 7
- 7.46% or 31.34 perf counts in [the earlier baseline](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_2_X_2021-11-01-2300/slc7_amd64_gcc900/profiling/11834.21/igprofCPU_step3.19/43) 
- 1.05% or 4.02 perf counts [this PR tests](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_2_X_2021-11-02-1100/slc7_amd64_gcc900/profiling/11834.21/PR-0b0595/20187/igprofCPU_step3/355)

This is roughly expected from the early return after the depth check (7 depths are checked).

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by swagata87** [comment]
thank you Slava,
this will be very useful for EGM HLT also

**Comment by slava77** [comment]
> thank you Slava, this will be very useful for EGM HLT also

does this imply a backport?
If so, to which release?

**Comment by slava77** [comment]
> in IB igprof tests the perf counts in `EgammaHcalIsolation::getHcalSum` have reduced by about a factor of 7
> 
> * 7.46% or 31.34 perf counts in [the earlier baseline](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_2_X_2021-11-01-2300/slc7_amd64_gcc900/profiling/11834.21/igprofCPU_step3.19/43)
> 
> * 1.05% or 4.02 perf counts [this PR tests](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_2_X_2021-11-02-1100/slc7_amd64_gcc900/profiling/11834.21/PR-0b0595/20187/igprofCPU_step3/355)
> 
> 
> This is roughly expected from the early return after the depth check (7 depths are checked).

the last update now has 
- 0.91% or 3.53 perf counts [last profiling as of 
f33bdaa ](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_2_X_2021-11-02-1100/slc7_amd64_gcc900/profiling/11834.21/PR-0b0595/20209/igprofCPU_step3/393)

So, slightly more than a factor of 8 speedup.

I think that further speedup is possible by refactoring the calling code. I added #35969 to track further progress with this still somewhat heavy code.

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/35955	f33bdaa
- code changes are technical, as confirmed by the PR comparisons/tests

**Comment by perrotta** [comment]
+1

**Comment by swagata87** [comment]
> does this imply a backport? If so, to which release?

no, I don't think a backport is needed

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc [hunk 1]
```diff
@@ -118,52 +118,60 @@ EgammaHcalIsolation::EgammaHcalIsolation(InclusionRule extIncRule,
...
   if (!right_depth)
     return 0.;
 
+  const bool goodHBe = hid.subdet() == HcalBarrel and hit.energy() > eThresHB_[h1];
+  const bool goodHEe = hid.subdet() == HcalEndcap and hit.energy() > eThresHE_[h1];
+  if (!(goodHBe or goodHEe))
+    return 0.;
+
+  const auto phit = caloGeometry_.getPosition(hit.detid());
+  const float phitEta = phit.eta();
```
Comment (VinInn) [inline RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: use cashed eta in caloGeometry_

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc [hunk 2]
```diff
@@ -118,52 +118,60 @@ EgammaHcalIsolation::EgammaHcalIsolation(InclusionRule extIncRule,
...
+
+  if (extIncRule_ == InclusionRule::withinConeAroundCluster or intIncRule_ == InclusionRule::withinConeAroundCluster) {
+    auto const dR2 = deltaR2(pcluEta, pcluPhi, phitEta, phit.phi());
```
Comment (VinInn) [inline RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:157]: use cashed phi in caloGeometry

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc (hunk 1 above)
Comment (slava77) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: please clarify the member method
Geometry/CaloGeometry/interface/CaloGeometry.h has only xyz position

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: RhoEtaPhi const& repPos() const { return m_rep; }
  virtual float rhoPos() const { return m_rep.rho(); }
  virtual float etaPos() const { return m_rep.eta(); }
  virtual float phiPos() const { return m_rep.phi(); }

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: eventually through CaloGeometry::getGeometry if not in the same SubDetector

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: one could extend the CaloGeometry interface to support ```repPos``` (copy paste code of ```getPosition```)
In my opinion would be more efficient to check first if the CellGeometry is the same and avoid the overhead of the code  (not inlined) to search for the subDetector for each single hit....

Comment (slava77) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: ah, so the suggestion is for an extension to the geometry interface; I thought at first that this is something that exists already.

IIRC, there is one rechit per cell in ECAL/HCAL

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: it exits already in CaloCellGeometry that is what CaloGeometry access. please have a look to the implementation in
https://cmssdt.cern.ch/dxr/CMSSW/source/Geometry/CaloGeometry/src/CaloGeometry.cc#50

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: here https://github.com/slava77/cmssw/blob/f33bdaa8be9c4aa4a3f336dbf23501b059a8988b/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L188
thre is a loop on the hits in HBorHE. one can verify if a hit is in the same SubDet of the previous and avoid all the search logic in CaloGeometry and acess directly CaloCellGeometry

Comment (VinInn) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:154]: here locally one can invoke
caloGeometry_.getGeometry(hit.detid())->repPos();

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc [hunk 3]
```diff
@@ -118,52 +118,60 @@ EgammaHcalIsolation::EgammaHcalIsolation(InclusionRule extIncRule,
...
   const bool right_depth = (depth == 0 or hd == depth);
   if (!right_depth)
     return 0.;
 
+  const bool goodHBe = hid.subdet() == HcalBarrel and hit.energy() > eThresHB_[h1];
+  const bool goodHEe = hid.subdet() == HcalEndcap and hit.energy() > eThresHE_[h1];
+  if (!(goodHBe or goodHEe))
+    return 0.;
+
+  const auto phit = caloGeometry_.getPosition(hit.detid());
```
Comment (VinInn) [inline RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:153]: suggested change:
```
const auto phit = caloGeometry_.getGeometry(hit.detid())->repPos();
```

Comment (slava77) [reply RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc:153]: > suggested change:

Thanks. Testing. 
PR soon.

**Pull request merged**